src/window/winit_window.rs

```diff
@@ -221,75 +231,106 @@ impl Window {
-
-                        self.window.set_inner_size(dpi::LogicalSize {
-                            width: browser_window.inner_width().unwrap().as_f64().unwrap(),
-                            height: browser_window.inner_height().unwrap().as_f64().unwrap(),
-                        });
+            }
+            Event::AboutToWait => {
+                self.window.request_redraw();
+            }
+            Event::WindowEvent { ref event, .. } => {
+                frame_input_generator.handle_winit_window_event(event);
+                match event {
+                    WindowEvent::Resized(physical_size) => {
+                        #[cfg(target_arch = "wasm32")]
+                        {
+                            use winit::platform::web::WindowExtWebSys;
+                            let canvas = self.window.canvas().unwrap();
+                            canvas.set_width(canvas.client_width() as u32);
+                            canvas.set_height(canvas.client_height() as u32);
+                        }
+                        self.gl.resize(*physical_size);
                     }
+                    WindowEvent::RedrawRequested => {
+                        #[cfg(target_arch = "wasm32")]
+                        if self.maximized || option_env!("THREE_D_SCREENSHOT").is_some() {
+                            use winit::platform::web::WindowExtWebSys;
+
+                            if let Some(html_canvas) = self.window.canvas() {
+                                let browser_window = html_canvas
+                                    .owner_document()
+                                    .and_then(|doc| doc.default_view())
+                                    .or_else(web_sys::window)
+                                    .unwrap();
+                                _ = self.window.request_inner_size(dpi::LogicalSize {
+                                    width: browser_window.inner_width().unwrap().as_f64().unwrap(),
+                                    height: browser_window
+                                        .inner_height()
+                                        .unwrap()
+                                        .as_f64()
+                                        .unwrap(),
+                                });
+                            }
+                        }
 
-                    let frame_input = frame_input_generator.generate(&self.gl);
-                    let frame_output = callback(frame_input);
-                    if frame_output.exit {
-                        *control_flow = ControlFlow::Exit;
-                    } else {
-                        if frame_output.swap_buffers && option_env!("THREE_D_SCREENSHOT").is_none()
+                        #[cfg(target_arch = "wasm32")]
                         {
-                            self.gl.swap_buffers().unwrap();
+                            // NOTE: This is needed because the canvas resize events are not all received.
```

Comment:
That's unfortunate that winit events are not working correctly 😞 But I think the fix is good then. However, I think it should just always make a resized event in each frame and then the frame input generator should check whether it actually has changed. That way, you don't need the `window_width` and `window_height` functionality on the frame input generator.